cmd/cloud/installation.go

```diff
@@ -423,3 +433,146 @@ var installationShowStateReport = &cobra.Command{
+			return errors.New("installation does not exist")
+		}
+		if installation.State != model.InstallationStateDeleted {
+			return errors.New("installation recovery can only be performed on deleted installations")
+		}
+
+		// DNS could have been claimed by a new installation, so we need to check
+		// that as well.
+		installations, err := sqlStore.GetInstallations(&model.InstallationFilter{
+			DNS:    installation.DNS,
+			Paging: model.AllPagesNotDeleted(),
+		}, false, false)
+		if err != nil {
+			return errors.Wrap(err, "failed to get cluster installations filtered by DNS")
+		}
+		if len(installations) != 0 {
+			return errors.Errorf("the requested installation's DNS is now in use by %d installations", len(installations))
+		}
+
+		// Be extra cautious until we can test other configs.
+		if installation.Database != model.InstallationDatabaseMultiTenantRDSPostgres {
+			return errors.Errorf("installation database type %s is not supported", installation.Database)
+		}
+		if installation.Filestore != model.InstallationFilestoreBifrost && installation.Filestore != model.InstallationFilestoreMultiTenantAwsS3 {
+			return errors.Errorf("installation filestore type %s is not supported", installation.Filestore)
+		}
+
+		clusterInstallations, err := sqlStore.GetClusterInstallations(&model.ClusterInstallationFilter{
+			InstallationID: installation.ID,
+			Paging:         model.AllPagesWithDeleted(),
+		})
+		if err != nil {
+			return errors.Wrap(err, "failed to get cluster installations")
+		}
+		if len(clusterInstallations) != 1 {
+			return errors.Errorf("expected to find 1 cluster installation, but found %d", len(clusterInstallations))
+		}
+		clusterInstallation := clusterInstallations[0]
+
+		db, err := sqlStore.GetMultitenantDatabase(databaseID)
+		if err != nil {
+			return errors.Wrap(err, "failed to find database")
+		}
+		if db == nil {
+			return errors.New("failed to find multitenant database")
+		}
+
+		logger.Info("Restoring AWS resources")
+
+		awsRegion := os.Getenv("AWS_REGION")
+		if awsRegion == "" {
+			awsRegion = toolsAWS.DefaultAWSRegion
+		}
+		awsConfig := &sdkAWS.Config{
+			Region:     sdkAWS.String(awsRegion),
+			MaxRetries: sdkAWS.Int(toolsAWS.DefaultAWSClientRetries),
+		}
+		awsClient, err := toolsAWS.NewAWSClientWithConfig(awsConfig, logger)
+		if err != nil {
+			return errors.Wrap(err, "failed to build AWS client")
+		}
+
+		err = awsClient.SecretsManagerRestoreSecret(toolsAWS.RDSMultitenantSecretName(installation.ID), logger)
+		if err != nil {
+			return errors.Wrap(err, "failed to recover AWS database secret")
+		}
+
+		logger.Info("Updating multitenant database")
+
+		locked, err := sqlStore.LockMultitenantDatabase(db.ID, instanceID)
+		if err != nil {
+			return errors.Wrap(err, "failed to lock multitenant database")
+		}
+		if !locked {
+			return errors.New("failed to acquire lock on multitenant database")
+		}
+		defer sqlStore.UnlockMultitenantDatabase(db.ID, instanceID, false)
```

Comment:
It is rather unlikely but unlocking can fail, so it would be good to log some warning in that case.

Reply:
Ah, thanks for reminding me. I had meant to update that part.